tasks/WebRTCTask.cpp

```diff
@@ -0,0 +1,449 @@
+    }
+    catch(const std::logic_error& e)
+    {
+        tmpPc.reset();
+        LOG_ERROR_S << e.what();
+        LOG_ERROR_S << std::endl;
+        // Json::Value msg;
+        // msg["type"] = "offer";
+        // msg["error"] = e.what();
+        // if (auto ws = wws.lock())
+		// 	ws->send(fast.write(msg));
+    }
+}
+
+void WebRTCTask::onAnswer(shared_ptr<rtc::WebSocket> wws) {
+    if (!this->getIdFromMessage(tmpPeerID)) {
+        return;
+    }
+
+    if (tmpPeerID != _peer_id.get()) {
+        return;
+    }
+
+    std::string peerId;
+    if (!this->getPeerIdFromMessage(peerId)) {
+        return;
+    }
+
+    if (peerId != _local_id.get()) {
+        return;
+    }
+
+    std::string answer;
+    if (!this->getDescriptionFromMessage(answer)) {
+        return;
+    }
+
+    try
+    {
+        tmpPc->setRemoteDescription(rtc::Description(answer, "answer"));
+    }
+    catch(const std::logic_error& e)
+    {
+        tmpPc.reset();
+        LOG_ERROR_S << e.what();
+        LOG_ERROR_S << std::endl;
+        // Json::Value msg;
+        // msg["type"] = "answer";
+        // msg["error"] = e.what();
+        // if (auto ws = wws.lock())
+		// 	ws->send(fast.write(msg));
+    }
+}
+
+void WebRTCTask::onCandidate(shared_ptr<rtc::WebSocket> wws) {
+    if (!tmpPc) {
+        // Json::Value msg;
+        // msg["type"] = "candidate";
+        // msg["error"] = "must send offer before candidates";
+        // if (auto ws = wws.lock())
+		// 	ws->send(fast.write(msg));
+        return;
+    }
+
+    std::string id;
+    if (!this->getIdFromMessage(id)) {
+        return;
+    }
+
+    if (id != tmpPeerID) {
+        return;
+    }
+
+    std::string peerId;
+    if (!this->getPeerIdFromMessage(peerId)) {
+        return;
+    }
+
+    if (peerId != _local_id.get()) {
+        return;
+    }
+
+    std::string candidate;
+    if (!this->getCandidateFromMessage(candidate)) {
+        return;
+    }
+
+    std::string mid;
+    if (!this->getMidFromMessage(mid)) {
+        return;
+    }
+
+    try
+    {
+        tmpPc->addRemoteCandidate(rtc::Candidate(candidate, mid));
+    }
+    catch(const std::logic_error& e)
+    {
+        LOG_ERROR_S << e.what();
+        LOG_ERROR_S << std::endl;
+        // Json::Value msg;
+        // msg["type"] = "candidate";
+        // msg["error"] = e.what();
+        // if (auto ws = wws.lock())
+		// 	ws->send(fast.write(msg));
+    }
+}
+
+bool WebRTCTask::parseIncomingMessage(char const* data) {
+    std::string errs;
+    if (!decoder->parseJSONMessage(data, errs)) {
+        LOG_ERROR_S << "Failed parsing the message, got error: " << errs << std::endl;
+        return false;
+    }
+    return true;
+}
+
+bool WebRTCTask::getTypeFromMessage(std::string &out_str) {
+    if (!decoder->validateType()) {
+        return false;
+    }
+    out_str = decoder->getType();
+    return true;
+}
+
+bool WebRTCTask::getIdFromMessage(std::string &out_str) {
```

Comment:
Do not use out arguments (please have a read on our coding standards)

Reply:
yeah, sure. When we are creating from scratch is easier to keeps the pattern. How I'm coding following the lib example, indeed there are a lot of changes to do.